cpu/sam0_common/periph/adc.c

```diff
@@ -63,20 +83,106 @@ static void _adc_poweroff(void)
 {
     while (_adc_syncing()) {}
     /* Disable */
-    ADC->CTRLA.reg &= ~ADC_CTRLA_ENABLE;
+    ADC_DEV->CTRLA.reg &= ~ADC_CTRLA_ENABLE;
     while (_adc_syncing()) {}
     /* Disable bandgap */
 #ifdef CPU_SAMD21
     if (ADC_REF_DEFAULT == ADC_REFCTRL_REFSEL_INT1V) {
         SYSCTRL->VREF.reg &= ~SYSCTRL_VREF_BGOUTEN;
     }
-#else /* CPU_SAML21 */
+#else
     if (ADC_REF_DEFAULT == ADC_REFCTRL_REFSEL_INTREF) {
         SUPC->VREF.reg &= ~SUPC_VREF_VREFOE;
     }
 #endif
 }
 
+static void _setup_clock(void)
+{
+    /* Enable gclk in case we are the only user */
+    sam0_gclk_enable(ADC_GCLK_SRC);
+#ifdef CPU_SAMD21
+    /* Power On */
+    PM->APBCMASK.reg |= PM_APBCMASK_ADC;
+    /* GCLK Setup */
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN
+            | GCLK_CLKCTRL_GEN(ADC_GCLK_SRC) | (GCLK_CLKCTRL_ID(ADC_GCLK_ID)));
+    /* Configure prescaler */
+    ADC_DEV->CTRLB.reg = ADC_PRESCALER;
+#else
+    /* Power on */
+    #ifdef MCLK_APBCMASK_ADC
+        MCLK->APBCMASK.reg |= MCLK_APBCMASK_ADC;
+    #else
+        #ifdef MCLK_APBDMASK_ADC0
+            if (ADC_DEV == ADC0) {
+                MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC0;
+            } else {
+                MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC1;
+            }
+        #else
+            MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC;
+        #endif
+    #endif
+
+    #ifdef ADC0_GCLK_ID
+        /* GCLK Setup */
+        if (ADC_DEV == ADC0) {
+            GCLK->PCHCTRL[ADC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
+                    | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+        }
+        else {
+            GCLK->PCHCTRL[ADC1_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
+                    | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+        }
+        /* Configure prescaler */
+        ADC_DEV->CTRLA.reg = ADC_PRESCALER;
+    #else
+        /* GCLK Setup */
+        GCLK->PCHCTRL[ADC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
+                | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+        /* Configure prescaler */
+        ADC_DEV->CTRLB.reg = ADC_PRESCALER;
+    #endif
+#endif
+}
+
+static void _setup_calibration(void)
+{
+#ifdef CPU_SAMD21
+    /* Load the fixed device calibration constants */
+    ADC_DEV->CALIB.reg =
+        ADC_CALIB_BIAS_CAL((*(uint32_t*)ADC_FUSES_BIASCAL_ADDR >>
+                            ADC_FUSES_BIASCAL_Pos)) |
+        ADC_CALIB_LINEARITY_CAL((*(uint64_t*)ADC_FUSES_LINEARITY_0_ADDR >>
+                                ADC_FUSES_LINEARITY_0_Pos));
+#else
+    /* Set default calibration from NVM */
+    #ifdef ADC0_FUSES_BIASCOMP_ADDR
+        if (ADC_DEV == ADC0) {
+            ADC_DEV->CALIB.reg =
+                ADC0_FUSES_BIASCOMP((*(uint32_t*)ADC0_FUSES_BIASCOMP_ADDR)) >>
+                ADC_CALIB_BIASCOMP_Pos |
```

Comment:
Shouldn't that be
```suggestion
                ADC0_FUSES_BIASCOMP((*(uint32_t*)ADC0_FUSES_BIASCOMP_ADDR)) << ADC_CALIB_BIASCOMP_Pos
```

(Also keep the shift on a single line to avoid confusion. 80 Characters are just a *recommendation*.

Reply:
Ah this is to 're-normalize' the fuse value 

Looking closer this evaluates to 

```C
(ADC_FUSES_BIASCOMP_Msk & ((value) << ADC_FUSES_BIASCOMP_Pos)) >>
                ADC_CALIB_BIASCOMP_Pos
```

So while `ADC0_FUSES_BIASCOMP_ADDR & 0x7` would be shorter, using the macros keeps it future proof. (and the compiler will optimize it anyway)